### DISCUSSION TITLE
Once the application is built and signed, it does not install

### DISCUSSION THREAD

**Discussion by YaguriDev** [iBotPeaches/Apktool#3279] [Q&A] [answered]
Java 8 \ JDK 8
As the name implies, I disassembled the `./apktool d app.apk` and then reassembled without making any changes `./apktool b app`.

I signed the app using **SignApk**. The app is not installed on the device, giving the error `App not installed`. 

On the emulator (_BlueStacks_) the application functions.
What could be the reason? Is it the signature or the app build and version of the app?

~Thanks in advance to whoever can help sort out this problem.

**Comment by iBotPeaches** [comment]
>  What could be the reason? Is it the signature or the app build and version of the app?

Tough to tell. Normally you'll get the reason printed into the logcat, so you can with `adb logcat` while the install is going on.

**Reply by YaguriDev** [reply]
I did this one and got this error:
```
Scanning Failed.
PackageManager: com.android.server.pm.PackageManagerException: No signature found in package of version 2 or newer for package com.<application>
```

After googling, I found a solution: use zipalign. I used zipalign both before and after signing the application - nothing changed and the error remained. Next, I decided to sign the application with my own key via apksigner. The result is the same.

jarsigner:

`jarsigner -verbose -keystore <keystore>.keystore <apk>.apk <alias>`
I also tried:
`jarsigner -verbose -sigalg SHA1withRSA -digestalg SHA1 -keystore <keystore>.keystore <apk>.apk <alias>`

zipalign:
`zipalign -p -f -v 4 <input>.apk <output>.apk`

Tried a lot of things, but still haven't found a solution.

**Reply by YaguriDev** [reply, marked as answer]
I fixed it. 
First I installed jdk11, then I created a key using that jdk and keytool I created the key.
`./keytool -genkeypair -v -keystore <keystore>.keystore -alias <alias> -keyalg RSA -keysize 2048 -validity 10000`
Next, using zipalign from the SDK, I zipped the APK file.
`./zipalign -f -v 4 <app>.apk appzip.apk`
After that, using apksigner from the same SDK, I signed the app.
`./apksigner sign --ks <keystore>.keystore --ks-key-alias <alias> appzip.apk`

Now everything works. It took me about 4 days. Most likely the problem was in the JDK version, or in the SDK version. The main thing is that everything works now.

**Reply by iBotPeaches** [reply]
Glad you got it. Attaching this link for future googlers - https://developer.android.com/tools/zipalign
